File: 2015/Haskell/Day1.hs
part1 :: IO Int
part1 = do
    file <- readFile "Day1Input.txt"
    let final_floor = foldl (\acc inst -> if inst == '(' then acc + 1 else acc - 1) 0 file
    print "Answer to part 1:"
    return final_floor

part2 :: IO Int
part2 = do
    file <- readFile "Day1Input.txt"
    let floors = scanl (\acc inst -> if inst == '(' then acc + 1 else acc - 1) 0 file
    let answer_part2 = find_basement floors
    print "Answer to part 2:"
    return answer_part2

    where find_basement :: [Int] -> Int
          find_basement (h:t)
              | h == -1 = 0
              | otherwise = 1 + find_basement t

main :: IO Int
main = do
    answer1 <- part1
    print answer1
    part2

File: 2015/Python/Day1.py
def get_input():
    with open("Day1Input.txt") as file:
        char = file.read(1)
        while char:
            yield char
            char = file.read(1)

def main():
    floor = 0
    answer_part2 = 0
    for pos, inst in enumerate(get_input(),1):
        floor += 1 if inst == "(" else -1
        if floor == -1 and not answer_part2:
            answer_part2 = pos
    print(f"Answer to part 1: {floor}")
    print(f"Answer to part 2: {answer_part2}")

if __name__ == "__main__":
    main()